packages/aws-library/src/aws_library/s3/_client.py

```diff
@@ -0,0 +1,460 @@
+
+    @s3_exception_handler(_logger)
+    async def create_single_presigned_upload_link(
+        self, *, bucket: S3BucketName, object_key: S3ObjectKey, expiration_secs: int
+    ) -> AnyUrl:
+        # NOTE: ensure the bucket/object exists, this will raise if not
+        await self._client.head_bucket(Bucket=bucket)
+        generated_link = await self._client.generate_presigned_url(
+            "put_object",
+            Params={"Bucket": bucket, "Key": object_key},
+            ExpiresIn=expiration_secs,
+        )
+        url: AnyUrl = parse_obj_as(AnyUrl, generated_link)
+        return url
+
+    @s3_exception_handler(_logger)
+    async def create_multipart_upload_links(
+        self,
+        *,
+        bucket: S3BucketName,
+        object_key: S3ObjectKey,
+        file_size: ByteSize,
+        expiration_secs: int,
+        sha256_checksum: SHA256Str | None,
+    ) -> MultiPartUploadLinks:
+        # NOTE: ensure the bucket exists, this will raise if not
+        await self._client.head_bucket(Bucket=bucket)
+        # first initiate the multipart upload
+        create_input: dict[str, Any] = {"Bucket": bucket, "Key": object_key}
+        if sha256_checksum:
+            create_input["Metadata"] = {"sha256_checksum": sha256_checksum}
+        response = await self._client.create_multipart_upload(**create_input)
+        upload_id = response["UploadId"]
+        # compute the number of links, based on the announced file size
+        num_upload_links, chunk_size = compute_num_file_chunks(file_size)
+        # now create the links
+        upload_links = parse_obj_as(
+            list[AnyUrl],
+            await asyncio.gather(
+                *(
+                    self._client.generate_presigned_url(
+                        "upload_part",
+                        Params={
+                            "Bucket": bucket,
+                            "Key": object_key,
+                            "PartNumber": i + 1,
+                            "UploadId": upload_id,
+                        },
+                        ExpiresIn=expiration_secs,
+                    )
+                    for i in range(num_upload_links)
+                ),
+            ),
+        )
+        return MultiPartUploadLinks(
+            upload_id=upload_id, chunk_size=chunk_size, urls=upload_links
+        )
+
+    @s3_exception_handler(_logger)
+    async def list_ongoing_multipart_uploads(
+        self,
+        *,
+        bucket: S3BucketName,
+    ) -> list[tuple[UploadID, S3ObjectKey]]:
+        """Returns all the currently ongoing multipart uploads
+
+        NOTE: minio does not implement the same behaviour as AWS here and will
+        only return the uploads if a prefix or object name is given [minio issue](https://github.com/minio/minio/issues/7632).
+
+        :return: list of AWS uploads see [boto3 documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.list_multipart_uploads)
```

Comment:
Q: we still care about minio because it used locally? Or we are using the in memory implementation locally?

Not sure if this is still relevant or up to date

Reply:
good that you checked that comment. I might have removed the key stuff. will check!

Reply:
I will keep it there for now.